internal/k8s/controllers/http_route_controller.go

```diff
@@ -53,5 +60,126 @@ func (r *HTTPRouteReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 func (r *HTTPRouteReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&gateway.HTTPRoute{}).
+		Watches(
+			&source.Kind{Type: &gateway.ReferencePolicy{}},
+			handler.EnqueueRequestsFromMapFunc(r.referencePolicyToRouteRequests),
+		).
 		Complete(gatewayclient.NewRequeueingMiddleware(r.Log, r))
 }
+
+// For UpdateEvents which contain both a new and old object, this transformation
+// function is run on both objects and both sets of Requests are enqueued.
+//
+// This is needed to reconcile any objects matched by  both current and prior
+// state in case a ReferencePolicy has been modified to revoke permission from a
+// namespace or to a service
+//
+// It may be possible to improve performance here by filtering Routes by
+// BackendRefs selectable by the To fields, but currently we just revalidate
+// all Routes allowed in the From Namespaces
+func (r *HTTPRouteReconciler) referencePolicyToRouteRequests(object client.Object) []reconcile.Request {
+	// TODO: Is there a safer way I could typecheck this with
+	// object.GetObjectKind() or something before casting?
+	refPolicy := *object.(*gateway.ReferencePolicy)
```

Comment:
this is how you'd do the safe typecast, but you may not need to given the fact that the `Watches` ensures the proper typing.

```suggestion
	refPolicy, ok := object.(*gateway.ReferencePolicy)
	if !ok {
		return nil
	}
```

Also, is there a reason for dereferencing after the cast?

Reply:
The deref is for the following call to `r.getRoutesAffectedByReferencePolicy(refPolicy)` (could deref at call site instead I guess)

Reply:
Pretty sure this is guaranteed never to be `nil` with how the controller runtime calls things (and if it was `nil`, this could potentially panic with a `nil` pointer dereference), but the only real use comes from iterating over `refPolicy.Spec.From` which shouldn't require the dereference. I'd just drop it and change the function signatures to take a pointer.